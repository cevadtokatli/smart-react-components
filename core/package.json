{
  "name": "@smart-react-components/core",
  "version": "2.0.0-beta-0",
  "license": "MIT",
  "main": "index.js",
  "description": "SRC core includes the theme, Styled properties, and Element components.",
  "keywords": [
    "javascript",
    "typescript",
    "react",
    "styled-components"
  ],
  "author": {
    "name": "Cevad Tokatlı",
    "email": "cevadtokatli@hotmail.com"
  },
  "contributors": [
    {
      "name": "Cemil Tokatlı",
      "email": "cemil@tokatli.me"
    }
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/cevadtokatli/smart-react-components/tree/v2/core"
  },
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "yarn clear && tsc --outDir lib && yarn copy",
    "clear": "rm -rf lib",
    "copy": "cp package.json lib/package.json && cp README.md lib/README.md && cp ../LICENSE lib/LICENSE",
    "dev": "tsc -w --outDir lib",
    "lint": "eslint src",
    "release": "../scripts/release.sh core",
    "test": "jest"
  },
  "devDependencies": {
    "@types/color": "3.0.3",
    "@types/enzyme": "3.10.12",
    "@types/jest": "29.0.1",
    "@types/react": "18.0.19",
    "@types/react-dom": "18.0.6",
    "@types/styled-components": "5.1.26",
    "@wojtekmaj/enzyme-adapter-react-17": "0.6.7",
    "enzyme": "3.11.0",
    "eslint": "8.0.1",
    "jest": "29.0.3",
    "jsdom": "20.0.0",
    "jsdom-global": "3.0.2",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "styled-components": "5.3.5",
    "ts-jest": "29.0.0",
    "typescript": "4.8.3"
  },
  "dependencies": {
    "color": "4.2.3"
  },
  "peerDependencies": {
    "react": "^18.2.0",
    "styled-components": "^5.3.5"
  }
}
