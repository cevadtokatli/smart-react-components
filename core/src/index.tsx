export * from './theme'
export { default as createTheme } from './theme'
export * from './styled-props'
export * from './Element'
