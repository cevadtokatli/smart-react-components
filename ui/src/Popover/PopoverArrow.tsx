import React from 'react'
import styled from 'styled-components'
import { Position } from '../types'

export default styled.svg.attrs({
  children: (
    <>
      <path d="M251.8,371.6c-17.4,5.9-37.2,2.6-50.4-8L16.3,215.3v-0.1c-18.1-14.5-18.1-38.1,0-52.6L201.4,14.2 c13.3-10.7,33.1-13.8,50.4-8" />
      <path d="M371,0c5.9,17.4,2.6,37.2-8,50.4L214.7,235.5h-0.1c-14.5,18.1-38.1,18.1-52.6,0L13.6,50.4 C2.9,37.1-0.2,17.3,5.6,0" />
    </>
  ),
})`
  position: absolute;
  stroke-width: 20px;

  &[data-src-position="${Position.LEFT}"],
  &[data-src-position="${Position.RIGHT}"] {
    height: 15px;
    width: 10px;

    path:last-child {
      display: none;
    }
  }

  &[data-src-position="${Position.TOP}"],
  &[data-src-position="${Position.BOTTOM}"] {
    height: 10px;
    width: 15px;

    path:first-child {
      display: none;
    }
  }

  &[data-src-position="${Position.LEFT}"],
  &[data-src-position="${Position.BOTTOM}"] {
    transform: rotate(180deg);
  }
`
