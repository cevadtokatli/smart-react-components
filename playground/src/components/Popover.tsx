import { Div } from '@smart-react-components/core'
import Popover from '@smart-react-components/ui/Popover'
import { Position } from '@smart-react-components/ui/types'
import React from 'react'

export const Palette = () => (
  <Div marginTop={125}>
    <Popover position={Position.TOP} space={0}>
      <button style={{ marginLeft: 20 }}>Dynamic</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="!dynamic" position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>!Dynamic</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="primary" position={Position.RIGHT} space={0}>
      <button style={{ marginLeft: 20 }}>Primary</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="secondary" position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Secondary</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="success" position={Position.LEFT} space={0}>
      <button style={{ marginLeft: 20 }}>Success</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="danger" position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Danger</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="warning" position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Warning</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="info" position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Info</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
  </Div>
)

export const Soft = () => (
  <Div marginTop={125}>
    <Popover isSoft position={Position.TOP} space={0}>
      <button style={{ marginLeft: 20 }}>Dynamic</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="!dynamic" isSoft position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>!Dynamic</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="primary" isSoft position={Position.RIGHT} space={0}>
      <button style={{ marginLeft: 20 }}>Primary</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="secondary" isSoft position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Secondary</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="success" isSoft position={Position.LEFT} space={0}>
      <button style={{ marginLeft: 20 }}>Success</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="danger" isSoft position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Danger</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="warning" isSoft position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Warning</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover palette="info" isSoft position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>Info</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
  </Div>
)

export const Size = () => (
  <Div marginTop={125}>
    <Popover size="small" position={Position.TOP} space={0}>
      <button style={{ marginLeft: 20 }}>Dynamic</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover position={Position.RIGHT} space={5}>
      <button style={{ marginLeft: 20 }}>!Dynamic</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
    <Popover size="large" position={Position.RIGHT} space={0}>
      <button style={{ marginLeft: 20 }}>Primary</button>
      <span>Header</span>
      <div>Popover</div>
    </Popover>
  </Div>
)
