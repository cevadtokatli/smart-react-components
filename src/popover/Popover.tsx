import React from "react"
import {findDOMNode, createPortal} from "react-dom"
import {SizeProps} from "../props/size-props"
import {JSXElementProps, SetValue} from "../props"
import {TooltipAttributes} from "../types/tooltip"
import DV from "../default-value"
import DOMHelper from "../helper/DOMHelper"
import TooltipHelper from "../helper/TooltipHelper"
import PopoverElement from "./Element"
import ArrowElement from "../tooltip/ArrowElement"

export interface Props extends SizeProps {
    children: [JSX.Element, JSX.Element]
    elementProps?: JSXElementProps
    status: boolean
    setStatus: SetValue<boolean>
    type?: string
    position?: string
    dismissible?: boolean
    renderInBody?: boolean
}

interface State {
    attrs: TooltipAttributes
}

const DEFAULT_STYLE = {}

export default class Popover extends React.Component<Props, State> {
    private static defaultProps = {
        size: "default",
        elementProps: DV.JSX_ELEMENT_PROPS,
        type: "white",
        position: "right",
        renderInBody: true
    }

    private el = React.createRef<HTMLDivElement>()
    private arrowEl = React.createRef<HTMLDivElement>()

    constructor(props) {
        super(props)

        this.state = {
            attrs: {}
        }
    }

    componentDidMount() {
        const containerEl = findDOMNode(this) as HTMLElement
        
        if(!this.props.dismissible)
            DOMHelper.addEventListener(containerEl, ["click"], this.toggle)
        else {
            DOMHelper.addEventListener(containerEl, ["click"], this.open)
            if(this.props.status)
                DOMHelper.addEventListener(document.body, ["click"], this.close)
        }

        if(this.props.status)
            DOMHelper.addEventListener(window, ["resize", "scroll"], this.setPosition, true)

        this.setPosition()
    }

    componentDidUpdate(op) {
        const np = this.props

        if(np.status != op.status) {
            if(np.dismissible) {
                if(np.status)
                    DOMHelper.addEventListener(document.body, ["click"], this.close)
                else
                    DOMHelper.removeEventListener(document.body, ["click"], this.close)
            }
        
            if(np.status) {
                DOMHelper.addEventListener(window, ["resize", "scroll"], this.setPosition, true)
                this.setPosition()
            } else
                DOMHelper.removeEventListener(window, ["resize", "scroll"], this.setPosition)
        }
    }

    componentWillUnmount() {
        const containerEl = findDOMNode(this) as HTMLElement

        if(!this.props.dismissible)
            DOMHelper.removeEventListener(containerEl, ["click"], this.toggle)
        else {
            DOMHelper.removeEventListener(containerEl, ["click"], this.open)
            DOMHelper.removeEventListener(document.body, ["click"], this.close)
        }
        
        DOMHelper.removeEventListener(window, ["resize", "scroll"], this.setPosition)
    }

    /**
     * Toggles element.
     */
    toggle = () => this.props.setStatus(!this.props.status)

    /**
     * Open element.
     */
    open = () => {
        if(!this.props.status)
            this.props.setStatus(true)
    }

    /**
     * Closes element if target is not in the container element.
     */
    close = (e:Event) => {
        const containerEl = findDOMNode(this) as HTMLElement
        if(e.target != containerEl && !containerEl.contains(e.target as HTMLElement))
            this.props.setStatus(false)
    }

    /**
     * Sets position of the element.
     */
    setPosition = () => {
        if(this.props.status) {
            const containerEl = findDOMNode(this) as HTMLElement
            const headerEl = this.el.current.querySelector(`.src-popover-header`) as HTMLElement
            this.setState({
                attrs: TooltipHelper.getAttributes(this.el.current, containerEl, this.props.position, headerEl, this.arrowEl.current)
            })
        }
    }

    render() {
        const p = this.props
        const s = this.state

        const Element = (
            <PopoverElement maxWidth="300px" {...p.elementProps} style={{...(p.elementProps.style || DEFAULT_STYLE),transform:s.attrs.transform}} ref={this.el} size$={p.size} size$Sm={p.sizeSm} size$Md={p.sizeMd} size$Lg={p.sizeLg} size$Xl={p.sizeXl} type={p.type} arrow={s.attrs.arrow} isArrowInHeader={s.attrs.isArrowInHeader}>
                {p.children[1]}
                <ArrowElement zIndex="popover" ref={this.arrowEl} style={{transform:s.attrs.arrowTransform}} />
            </PopoverElement>
        )

        return (
            <>
                {p.children[0]}
                {this.props.status && (DOMHelper.canBeRenderedInPortal(p.renderInBody) ? createPortal(Element, document.body) : Element)}
            </>
        )
    }
}