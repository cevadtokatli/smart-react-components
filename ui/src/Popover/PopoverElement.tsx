import { StyledProps } from '@smart-react-components/core/styled-props'
import Div from '@smart-react-components/core/Element/Div'
import styled from 'styled-components'
import { PaletteProp, ResponsiveProp, SizeProp } from '@smart-react-components/core/types'
import PopArrow from '../components/PopArrow'
import { generateCSSTransitionClasses } from '../util/popover'

export const PopoverHeader = styled.h4`
  box-sizing: border-box;
  margin: 0;
`

export const PopoverContent = styled.div`
  box-sizing: border-box;
`

interface Props extends
  StyledProps,
  ResponsiveProp<'popoverSize', SizeProp> {
  duration: number
  hasTransition: boolean
  isSoft: boolean
  palette: PaletteProp
}

export default styled(Div).attrs({
  getPopoverSize: (v, t) => `
    > ${PopoverHeader} {
      font-size: ${t.$.size.popover[v].headerFontSize};
      padding: ${t.$.size.popover[v].headerPadding.y} ${t.$.size.popover[v].headerPadding.x};
    }

    > ${PopoverContent} {
      font-size: ${t.$.size.popover[v].fontSize};
      padding: ${t.$.size.popover[v].padding.y} ${t.$.size.popover[v].padding.x};
    }
  `,
})<Props>(({ theme, palette, duration, hasTransition, isSoft }: Props) => `
  box-sizing: border-box;
  color: ${theme.$.palette[palette].popover.font};
  fill: ${theme.$.palette[palette].popover.font};
  font-family: ${theme.$.fontFamily.popover};
  position: fixed;
  z-index: ${theme.$.zIndex.popover};

  ${PopoverContent} {
    background: ${theme.$.palette[palette].popover.background};
    border: solid 1px ${theme.$.palette[palette].popover.border};
  }

  ${PopoverHeader} {
    ${isSoft
      ? `
        background: ${theme.$.palette[palette].popover.softHeaderBackground};
        border: solid 1px ${theme.$.palette[palette].popover.softHeaderBorder};
        color: ${theme.$.palette[palette].popover.softHeaderFont};
        fill: ${theme.$.palette[palette].popover.softHeaderFont};

        ~ ${PopArrow}[data-arrow-header="true"] {
          fill: ${theme.$.palette[palette].popover.softHeaderBackground};
          stroke: ${theme.$.palette[palette].popover.softHeaderBorder};
        }
      `
      : `
        background: ${theme.$.palette[palette].popover.headerBackground};
        border: solid 1px ${theme.$.palette[palette].popover.headerBorder};
        color: ${theme.$.palette[palette].popover.headerFont};
        fill: ${theme.$.palette[palette].popover.headerFont};

        ~ ${PopArrow}[data-arrow-header="true"] {
          fill: ${theme.$.palette[palette].popover.headerBackground};
          stroke: ${theme.$.palette[palette].popover.headerBorder};
        }
      `
    }

    + ${PopoverContent} {
      border-top: 0;
    }
  }

  ${PopArrow}[data-arrow-header="false"] {
    fill: ${theme.$.palette[palette].popover.background};
    stroke: ${theme.$.palette[palette].popover.border};
  }

  ${hasTransition ? generateCSSTransitionClasses('popover', duration) : ''}
`)
