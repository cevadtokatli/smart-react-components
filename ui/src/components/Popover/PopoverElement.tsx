import { StyledProps } from '@smart-react-components/core/styled-props'
import Div from '@smart-react-components/core/Element/Div'
import styled from 'styled-components'
import { PaletteProp, ResponsiveProp, SizeProp } from '@smart-react-components/core/types'
import PopoverArrow from './PopoverArrow'
import { generateCSSTransitionClasses } from '../../util/popover'

export const PopoverHeader = styled.h4`
  box-sizing: border-box;
  margin: 0;
`

export const PopoverContent = styled.div`
  box-sizing: border-box;
`

interface Props extends
  StyledProps,
  ResponsiveProp<'popoverSize', SizeProp> {
  duration: number
  hasTransition: boolean
  isSoft: boolean
  palette: PaletteProp
}

export default styled(Div).attrs({
  getPopoverSize: (v, t) => `
    > ${PopoverHeader} {
      font-size: ${t.$.size.popover[v].headerFontSize};
      padding: ${t.$.size.popover[v].headerPadding.y} ${t.$.size.popover[v].headerPadding.x};
    }

    > ${PopoverContent} {
      font-size: ${t.$.size.popover[v].fontSize};
      padding: ${t.$.size.popover[v].padding.y} ${t.$.size.popover[v].padding.x};
    }
  `,
})<Props>(({ theme, palette, duration, hasTransition, isSoft }: Props) => `
  box-sizing: border-box;
  color: ${theme.$.color.dynamic.font};
  fill: ${theme.$.color.dynamic.font};
  font-family: ${theme.$.fontFamily.popover};
  position: fixed;
  z-index: ${theme.$.zIndex.popover};

  ${PopoverContent} {
    background: ${theme.$.color.dynamic.background};
    border: solid 1px ${theme.$.color.dynamic.accent};
  }

  ${PopoverHeader} {
    ${isSoft
      ? `
        background: ${theme.$.palette[palette].soft};
        border: solid 1px ${theme.$.palette[palette].softDynamicer};
        color: ${theme.$.palette[palette].softFont};
        fill: ${theme.$.palette[palette].softFont};

        ~ ${PopoverArrow}[data-arrow-header="true"] {
          fill: ${theme.$.palette[palette].soft};
          stroke: ${theme.$.palette[palette].softDynamicer};
        }
      `
      : `
        background: ${theme.$.palette[palette].main};
        border: solid 1px ${theme.$.palette[palette].dynamicer};
        color: ${theme.$.palette[palette].font};
        fill: ${theme.$.palette[palette].font};

        ~ ${PopoverArrow}[data-arrow-header="true"] {
          fill: ${theme.$.palette[palette].main};
          stroke: ${theme.$.palette[palette].dynamicer};
        }
      `
    }

    + ${PopoverContent} {
      border-top: 0;
    }
  }

  ${PopoverArrow}[data-arrow-header="false"] {
    fill: ${theme.$.color.dynamic.background};
    stroke: ${theme.$.color.dynamic.accent};
  }

  ${hasTransition ? generateCSSTransitionClasses('popover', duration) : ''}
`)
