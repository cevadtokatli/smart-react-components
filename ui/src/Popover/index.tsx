import { Theme } from '@smart-react-components/core/theme'
import { Partial, PaletteProp, ResponsiveProp, SetState, SizeProp } from '@smart-react-components/core/types'
import CSSTransition from '@smart-react-components/transition/CSSTransition'
import { TransitionAfterCallback, TransitionBeforeCallback } from '@smart-react-components/transition/types'
import React from 'react'
import { createPortal } from 'react-dom'
import { ThemeContext } from 'styled-components'
import PopArrow from '../components/PopArrow'
import useFixedBoxMethods from '../hooks/useFixedBoxMethods'
import { Position, TriggerInteraction } from '../types'
import { canBeRenderedInPortal } from '../util/dom'
import { calculatePosition, getPopArrowViewBox } from '../util/popover'
import PopoverElement, { PopoverHeader, PopoverContent } from './PopoverElement'

export { Position, TriggerInteraction } from '../types'

export interface Props extends Partial<ResponsiveProp<'size', SizeProp>> {
  afterHide?: TransitionAfterCallback
  afterShow?: TransitionAfterCallback
  beforeHide?: TransitionBeforeCallback
  beforeShow?: TransitionBeforeCallback
  children: [JSX.Element, JSX.Element] | [JSX.Element, JSX.Element, JSX.Element]
  hasHideAnimation?: boolean
  hasShowAnimation?: boolean
  isDismissible?: boolean
  isSoft?: boolean
  palette?: PaletteProp
  position?: Position
  setStatus?: SetState<boolean>
  space?: number
  status?: boolean
  transitionClassName?: string
  transitionDuration?: number
  triggerInteraction?: TriggerInteraction
}

const Popover: React.FC<Props> = ({
  afterHide,
  afterShow,
  beforeHide,
  beforeShow,
  children,
  hasHideAnimation = true,
  hasShowAnimation = true,
  isDismissible = true,
  isSoft,
  palette = 'dynamic',
  position = Position.RIGHT,
  setStatus,
  size = 'medium',
  sizeSm,
  sizeMd,
  sizeLg,
  sizeXl,
  space = 0,
  status,
  transitionClassName,
  transitionDuration,
  triggerInteraction = TriggerInteraction.CLICK,
}) => {
  const theme = React.useContext<Theme>(ThemeContext)
  const triggerEl = React.useRef<HTMLElement>(null)
  const boxEl = React.useRef<HTMLDivElement>(null)
  const headerEl = React.useRef<HTMLHeadingElement>(null)
  const arrowEl = React.useRef<SVGSVGElement>(null)

  const getTriggerEl = () => ((children[0] as any).ref ?? triggerEl).current as HTMLElement
  const getTransitionDuration = () => transitionDuration ?? theme.$.transition.popoverDuration

  const handlePosition = () => {
    const { arrowStyle, isArrowInHeader, pos, style } = calculatePosition(triggerEl.current.getBoundingClientRect(), boxEl.current.getBoundingClientRect(), headerEl.current?.getBoundingClientRect(), position, space)
    boxEl.current.setAttribute('style', style)
    boxEl.current.setAttribute('data-src-position', String(pos))
    arrowEl.current.setAttribute('style', arrowStyle)
    arrowEl.current.setAttribute('viewBox', getPopArrowViewBox(pos))
    arrowEl.current.setAttribute('data-src-position', String(pos))
    arrowEl.current.setAttribute('data-arrow-header', String(isArrowInHeader))
  }

  const {
    getStatus,
    handleBeforeShow,
  } = useFixedBoxMethods({ beforeShow, boxEl, getTriggerEl, handlePosition, isDismissible, key: 'popover', setStatus, status, triggerInteraction })

  const content = (
    <CSSTransition
      afterHide={afterHide}
      afterShow={afterShow}
      beforeHide={beforeHide}
      beforeShow={handleBeforeShow}
      className={transitionClassName ?? 'src-popover'}
      duration={getTransitionDuration()}
      hasHideAnimation={hasHideAnimation}
      hasShowAnimation={hasShowAnimation}
      status={getStatus()}
    >
      <PopoverElement
        duration={getTransitionDuration()}
        hasTransition={!transitionClassName}
        isSoft={isSoft}
        palette={palette}
        popoverSize={size}
        popoverSizeSm={sizeSm}
        popoverSizeMd={sizeMd}
        popoverSizeLg={sizeLg}
        popoverSizeXl={sizeXl}
        ref={boxEl}
      >
        { children.length === 3 && <PopoverHeader ref={headerEl}>{children[1]}</PopoverHeader> }
        { <PopoverContent>{children[2] ?? children[1]}</PopoverContent> }
        <PopArrow ref={arrowEl} />
      </PopoverElement>
    </CSSTransition>
  )

  return (
    <>
      { React.cloneElement(children[0], { ref: (children[0] as any).ref ?? triggerEl }) }
      { canBeRenderedInPortal() ? createPortal(content, document.body) : content }
    </>
  )
}

export default Popover
