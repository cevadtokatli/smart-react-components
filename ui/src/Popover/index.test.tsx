import { wrapTheme } from '@smart-react-components/core/test'
import { createEvent, fireEvent, render } from '@testing-library/react'
import React from 'react'
import Popover from './'

describe('<Popover />', () => {
  let children

  beforeAll(() => {
    children = [
      <button key="button">Button</button>,
      <div key="popover">Popover</div>,
      <span key="header">Header</span>,
    ]
  })

  it('should render component when status true', () => {
    const { asFragment } = render(wrapTheme(<Popover status>{children}</Popover>))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should not render component when status false', () => {
    const { asFragment } = render(wrapTheme(<Popover status={false}>{children}</Popover>))
    expect(asFragment()).toMatchSnapshot()
  })

  it('should call setStatus when clicked on button', () => {
    const setStatus = jest.fn()
    const { container } = render(wrapTheme(<Popover status={false} setStatus={setStatus}>{children}</Popover>))
    const node = container.querySelector('button')!
    fireEvent(node, createEvent.click(node))
    expect(setStatus).toHaveBeenCalled()
  })
})
