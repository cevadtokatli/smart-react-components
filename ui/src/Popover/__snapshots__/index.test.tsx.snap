// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popover /> should not render component when status false 1`] = `
<DocumentFragment>
  <button>
    Button
  </button>
</DocumentFragment>
`;

exports[`<Popover /> should render component when status true 1`] = `
<DocumentFragment>
  <button>
    Button
  </button>
  .c7 {
  position: absolute;
  stroke-width: 20px;
}

.c7[data-src-position="1"],
.c7[data-src-position="2"] {
  height: 15px;
  width: 10px;
}

.c7[data-src-position="1"] path:last-child,
.c7[data-src-position="2"] path:last-child {
  display: none;
}

.c7[data-src-position="4"],
.c7[data-src-position="8"] {
  height: 10px;
  width: 15px;
}

.c7[data-src-position="4"] path:first-child,
.c7[data-src-position="8"] path:first-child {
  display: none;
}

.c7[data-src-position="1"],
.c7[data-src-position="8"] {
  -webkit-transform: rotate(180deg);
  -ms-transform: rotate(180deg);
  transform: rotate(180deg);
}

.c0 > .c2 {
  font-size: 1rem;
  padding: .5rem .75rem;
}

.c0 > .c4 {
  font-size: .875rem;
  padding: .7rem .8rem;
}

.c3 {
  box-sizing: border-box;
  margin: 0;
}

.c5 {
  box-sizing: border-box;
}

.c1 {
  box-sizing: border-box;
  color: #323a46;
  fill: #323a46;
  font-family: inherit;
  position: fixed;
  z-index: 1000;
}

.c1 .c4 {
  background: #fff;
  border: solid 1px #dbdbdb;
}

.c1 .c2 {
  background: #f1f5f7;
  border: solid 1px #BBCED8;
  color: #323a46;
  fill: #323a46;
}

.c1 .c2 ~ .c6[data-arrow-header="true"] {
  fill: #f1f5f7;
  stroke: #BBCED8;
}

.c1 .c2 + .c4 {
  border-top: 0;
}

.c1 .c6[data-arrow-header="false"] {
  fill: #fff;
  stroke: #dbdbdb;
}

.c1.src-popover-show-active,
.c1.src-popover-hide-active {
  -webkit-transition: 300ms 0s ease-in-out;
  transition: 300ms 0s ease-in-out;
  -webkit-transition-property: -webkit-transform,opacity;
  -webkit-transition-property: transform,opacity;
  transition-property: transform,opacity;
}

.c1.src-popover-show {
  opacity: 0;
}

.c1.src-popover-show[data-src-position="1"] {
  -webkit-transform: translateX(-30px);
  -ms-transform: translateX(-30px);
  transform: translateX(-30px);
}

.c1.src-popover-show[data-src-position="2"] {
  -webkit-transform: translateX(30px);
  -ms-transform: translateX(30px);
  transform: translateX(30px);
}

.c1.src-popover-show[data-src-position="4"] {
  -webkit-transform: translateY(-30px);
  -ms-transform: translateY(-30px);
  transform: translateY(-30px);
}

.c1.src-popover-show[data-src-position="8"] {
  -webkit-transform: translateY(-30px);
  -ms-transform: translateY(-30px);
  transform: translateY(-30px);
}

.c1.src-popover-show-active {
  opacity: 1;
}

.c1.src-popover-show-active[data-src-position="1"],
.c1.src-popover-show-active[data-src-position="2"] {
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

.c1.src-popover-show-active[data-src-position="4"],
.c1.src-popover-show-active[data-src-position="8"] {
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c1.src-popover-hide {
  opacity: 1;
}

.c1.src-popover-hide[data-src-position="1"],
.c1.src-popover-hide[data-src-position="2"] {
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

.c1.src-popover-hide[data-src-position="4"],
.c1.src-popover-hide[data-src-position="8"] {
  -webkit-transform: translateY(0);
  -ms-transform: translateY(0);
  transform: translateY(0);
}

.c1.src-popover-hide-active {
  opacity: 0;
}

.c1.src-popover-hide-active[data-src-position="1"] {
  -webkit-transform: translateX(-30px);
  -ms-transform: translateX(-30px);
  transform: translateX(-30px);
}

.c1.src-popover-hide-active[data-src-position="2"] {
  -webkit-transform: translateX(30px);
  -ms-transform: translateX(30px);
  transform: translateX(30px);
}

.c1.src-popover-hide-active[data-src-position="4"] {
  -webkit-transform: translateY(-30px);
  -ms-transform: translateY(-30px);
  transform: translateY(-30px);
}

.c1.src-popover-hide-active[data-src-position="8"] {
  -webkit-transform: translateY(-30px);
  -ms-transform: translateY(-30px);
  transform: translateY(-30px);
}

<div
    class="c0 c1"
    data-src-position="8"
    style="
      left: 0px;
      top: 10px;
    "
  >
    <h4
      class="c2 c3"
    >
      <div>
        Popover
      </div>
    </h4>
    <div
      class="c4 c5"
    >
      <span>
        Header
      </span>
    </div>
    <svg
      class="c6 c7"
      data-arrow-header="true"
      data-src-position="8"
      style="
      left: -7.5px;
      top: -9px;
    "
      viewBox="0 0 377.3 251.9"
    >
      <path
        d="M251.8,371.6c-17.4,5.9-37.2,2.6-50.4-8L16.3,215.3v-0.1c-18.1-14.5-18.1-38.1,0-52.6L201.4,14.2 c13.3-10.7,33.1-13.8,50.4-8"
      />
      <path
        d="M371,0c5.9,17.4,2.6,37.2-8,50.4L214.7,235.5h-0.1c-14.5,18.1-38.1,18.1-52.6,0L13.6,50.4 C2.9,37.1-0.2,17.3,5.6,0"
      />
    </svg>
  </div>
</DocumentFragment>
`;
